File: frontend/src/lib/definitions/desktop.ts
import type { FilterFormData } from "@/lib/filter-utils";
import type {
  ChatResponse,
  CommitResponse,
  CommitsResponse,
  DatabaseResponse,
  FilterResponse,
  RepoDeleteResponse,
  RepoResponse,
} from "@/lib/definitions/api";
import type {
  ReviewCompareResponse,
  ReviewApprovalDecision,
  ReviewChatRequest,
  ReviewChatResponse,
  ReviewHistoryResponse,
  ReviewReport,
  ReviewRun,
  ReviewRuntimeEvent,
  ReviewSession,
} from "@/lib/definitions/review";
import type { User } from "@/lib/definitions/auth";

export type DesktopServiceState =
  | "running"
  | "starting"
  | "stopped"
  | "error"
  | "unavailable";

export interface DesktopServiceHealth {
  state: DesktopServiceState;
  message?: string;
  url?: string;
}

export type TargetKind = "managed_model" | "vertex_endpoint";
export type CapabilityName = "text_generation" | "embeddings" | "review";
export type ModelSource =
  | "managed_api_model"
  | "deployable_google_model"
  | "deployable_partner_model"
  | "vertex_endpoint"
  | "manual_resource_name";

export interface DesktopReviewSettings {
  pullRequestGuidelines: string;
}

export interface DesktopAdditionalReviewGuideline {
  id: string;
  text: string;
}

export interface DesktopAdditionalReviewGuidelineState {
  repoPath: string;
  draftGuideline: string;
  guidelines: DesktopAdditionalReviewGuideline[];
  updatedAt: string | null;
}

export interface DesktopAdditionalReviewGuidelineSaveInput {
  repoPath: string;
  draftGuideline: string;
  guidelines: DesktopAdditionalReviewGuideline[];
}

export interface DesktopRepoSettings {
  maxCommits: number;
  contextLines: number;
  pullRequestGuidelines: string;
}

export interface DesktopRepoSettingsSaveInput extends DesktopRepoSettings {
  repoPath: string;
}

export const DEFAULT_DESKTOP_REVIEW_SETTINGS: DesktopReviewSettings = {
  pullRequestGuidelines: "",
};

export const DEFAULT_DESKTOP_REPO_SETTINGS: DesktopRepoSettings = {
  maxCommits: 50,
  contextLines: 10,
  pullRequestGuidelines: "",
};

export interface GoogleAITarget {
  target_kind: TargetKind;
  resource_name: string;
  display_name: string;
  publisher?: string | null;
  version?: string | null;
  location?: string | null;
  capabilities: CapabilityName[];
  adapter_family?: string | null;
  embedding_output_dimension?: number | null;
  source?: ModelSource | null;
}

export interface AIRuntimeConfig {
  schema_version: number;
  google_project_id: string | null;
  google_location: string;
  capabilities: {
    text_generation: GoogleAITarget | null;
    embeddings: GoogleAITarget | null;
    review: GoogleAITarget | null;
  };
}

export interface DesktopAiSavedProfile {
  id: string;
  name: string;
  config: AIRuntimeConfig;
  secretValues: Record<string, string>;
  updatedAt: string;
}

export interface DesktopAiProfileSaveInput {
  id?: string | null;
  name: string;
  config: AIRuntimeConfig;
  secretValues: Record<string, string>;
}

export interface AICapabilityStatus {
  configured: boolean;
  ready: boolean;
  targetKind: TargetKind | null;
  resourceName: string | null;
  displayName: string | null;
  publisher: string | null;
  version: string | null;
  location: string | null;
  adapterFamily: string | null;
  embeddingOutputDimension?: number | null;
  message?: string;
  reindexRequired?: boolean;
}

export interface GoogleRuntimeStatus {
  projectId: string | null;
  location: string | null;
  adcReady: boolean;
  adcProjectId?: string | null;
  message?: string | null;
}

export interface DesktopSettingsStatus {
  firstRunCompleted: boolean;
  backendPort: number;
  dataDir: string;
  logDir: string;
  databaseUrlConfigured: boolean;
  aiRuntimeConfig: AIRuntimeConfig;
  savedAiProfiles: DesktopAiSavedProfile[];
  reviewSettings: DesktopReviewSettings;
  ai: {
    google: GoogleRuntimeStatus;
    textGeneration: AICapabilityStatus;
    embeddings: AICapabilityStatus;
    review: AICapabilityStatus;
  };
}

export interface DesktopHealthStatus {
  backend: DesktopServiceHealth;
  postgres: DesktopServiceHealth;
  authentication: {
    ready: boolean;
    desktopBackendReachable: boolean;
    desktopUserAvailable: boolean;
  };
  ai: {
    google: GoogleRuntimeStatus;
    textGeneration: AICapabilityStatus;
    embeddings: AICapabilityStatus;
    review: AICapabilityStatus;
  };
  desktopUser: {
    id: number;
    username: string;
    email?: string;
  } | null;
  credentials: {
    secretRefs: Record<string, boolean>;
  };
  settings: DesktopSettingsStatus;
}

export interface DesktopAiConfigInput {
  config: AIRuntimeConfig;
  secretValues: Record<string, string>;
}

export interface DesktopAiValidationResult {
  text_generation: {
    configured: boolean;
    ready: boolean;
    target_kind: TargetKind | null;
    resource_name: string | null;
    display_name: string | null;
    publisher: string | null;
    version: string | null;
    location: string | null;
    adapter_family: string | null;
    embedding_output_dimension?: number | null;
    message?: string | null;
  };
  embeddings: {
    configured: boolean;
    ready: boolean;
    target_kind: TargetKind | null;
    resource_name: string | null;
    display_name: string | null;
    publisher: string | null;
    version: string | null;
    location: string | null;
    adapter_family: string | null;
    embedding_output_dimension?: number | null;
    message?: string | null;
    reindex_required?: boolean;
  };
  review: {
    configured: boolean;
    ready: boolean;
    target_kind: TargetKind | null;
    resource_name: string | null;
    display_name: string | null;
    publisher: string | null;
    version: string | null;
    location: string | null;
    adapter_family: string | null;
    embedding_output_dimension?: number | null;
    message?: string | null;
  };
}

export interface GoogleModelGardenEntry {
  id: string;
  resource_name: string;
  display_name: string;
  publisher?: string | null;
  version?: string | null;
  location: string;
  target_kind: TargetKind;
  source: ModelSource;
  capabilities: CapabilityName[];
  adapter_family?: string | null;
  deployable: boolean;
  description?: string | null;
}

export interface GoogleModelGardenListResponse {
  items: GoogleModelGardenEntry[];
}

export interface GoogleTargetValidationInput {
  config: AIRuntimeConfig;
  capability: CapabilityName;
  target: GoogleAITarget;
}

export interface GoogleTargetValidationResult {
  configured: boolean;
  ready: boolean;
  capability: CapabilityName;
  target: GoogleAITarget;
  message?: string | null;
  embedding_output_dimension?: number | null;
}

export interface GoogleDeploymentInput {
  config: AIRuntimeConfig;
  model_resource_name: string;
  endpoint_resource_name: string;
  deployed_model_display_name: string;
  machine_type: string;
  accelerator_type?: string | null;
  accelerator_count?: number | null;
  min_replica_count?: number;
  max_replica_count?: number;
  accepted_terms: boolean;
  accepted_billing_notice: boolean;
}

export interface GoogleDeploymentResult {
  operation_name?: string | null;
  endpoint_resource_name: string;
  request: Record<string, unknown>;
  response: Record<string, unknown>;
}

export interface GitProjectSummary {
  path: string;
  name: string;
  lastOpenedAt: string;
}

export type RepoSyncPhase =
  | "planning"
  | "loading_commits"
  | "extracting_ast"
  | "embedding"
  | "writing_db"
  | "completed"
  | "failed";

export interface RepoSyncProgressEvent {
  progressId: string;
  repoPath: string;
  phase: RepoSyncPhase;
  label: string;
  percent: number;
  stagePercent: number;
  completedUnits: number;
  totalUnits: number;
  commitCount?: number | null;
  fileChangeCount?: number | null;
  hunkCount?: number | null;
  embeddingBatches?: number | null;
  insertedCommits?: number | null;
  error?: string | null;
  startedAt: string;
  updatedAt: string;
}

export interface GitOdysseyDesktopApi {
  getRepo(
    repoPath: string,
    repoSettings?: DesktopRepoSettings,
  ): Promise<RepoResponse>;
  ingestRepo(input: {
    repoPath: string;
    maxCommits: number;
    contextLines: number;
    force?: boolean;
  }): Promise<RepoResponse>;
  filterCommits(input: {
    query: string;
    filters: FilterFormData;
    repoPath: string;
    maxResults?: number;
  }): Promise<FilterResponse>;
  pickGitProject(): Promise<GitProjectSummary | null>;
  getRecentProjects(): Promise<GitProjectSummary[]>;
  getRepoSyncProgress(repoPath: string): Promise<RepoSyncProgressEvent | null>;
  deleteRepo(repoPath: string): Promise<RepoDeleteResponse>;
  summarizeCommit(sha: string): Promise<string>;
  summarizeFileChange(id: number): Promise<string>;
  summarizeHunk(id: number): Promise<string>;
  sendChatMessage(input: {
    query: string;
    repoPath: string;
    contextShas: string[];
    targetOverride?: GoogleAITarget | null;
  }): Promise<ChatResponse>;
  sendReviewChatMessage(input: ReviewChatRequest): Promise<ReviewChatResponse>;
  initDatabase(): Promise<DatabaseResponse>;
  dropDatabase(): Promise<DatabaseResponse>;
  getCommit(
    repoPath: string,
    commitSha: string,
    repoSettings?: DesktopRepoSettings,
  ): Promise<CommitResponse>;
  getCommits(
    repoPath: string,
    repoSettings?: DesktopRepoSettings,
  ): Promise<CommitsResponse>;
  compareReviewTarget(input: {
    repoPath: string;
    targetMode: "compare" | "commit";
    baseRef?: string;
    headRef?: string;
    commitSha?: string | null;
    contextLines: number;
  }): Promise<ReviewCompareResponse>;
  generateReview(input: {
    repoPath: string;
    targetMode: "compare" | "commit";
    baseRef?: string;
    headRef?: string;
    commitSha?: string | null;
    contextLines: number;
  }): Promise<ReviewReport>;
  createReviewSession(input: {
    repoPath: string;
    targetMode: "compare" | "commit";
    baseRef?: string;
    headRef?: string;
    commitSha?: string | null;
    contextLines: number;
  }): Promise<ReviewSession>;
  getReviewSession(sessionId: string): Promise<ReviewSession>;
  getReviewHistory(input: {
    repoPath: string;
    targetMode: "compare" | "commit";
    baseRef?: string;
    headRef?: string;
    commitSha?: string | null;
  }): Promise<ReviewHistoryResponse>;
  startReviewRun(input: {
    sessionId: string;
    customInstructions?: string | null;
  }): Promise<ReviewRun>;
  getReviewRun(input: { sessionId: string; runId: string }): Promise<ReviewRun>;
  cancelReviewRun(input: {
    sessionId: string;
    runId: string;
  }): Promise<ReviewRun>;
  respondReviewApproval(input: {
    sessionId: string;
    runId: string;
    approvalId: string;
    decision: ReviewApprovalDecision;
  }): Promise<ReviewRun>;
  getCurrentUser(): Promise<User>;
  logout(): Promise<{ message: string }>;
}

export interface GitOdysseyDesktopBridge {
  api: GitOdysseyDesktopApi;
  settings: {
    getStatus(): Promise<DesktopSettingsStatus>;
    getRepoSettings(repoPath: string): Promise<DesktopRepoSettings>;
    getAdditionalReviewGuidelines(
      repoPath: string,
    ): Promise<DesktopAdditionalReviewGuidelineState>;
    saveAdditionalReviewGuidelines(
      input: DesktopAdditionalReviewGuidelineSaveInput,
    ): Promise<DesktopAdditionalReviewGuidelineState>;
    saveReviewSettings(
      input: DesktopReviewSettings,
    ): Promise<DesktopReviewSettings>;
    validateAiConfig(
      input: DesktopAiConfigInput,
    ): Promise<DesktopAiValidationResult>;
    listGoogleModelGarden(input: {
      googleProjectId: string;
      googleLocation: string;
    }): Promise<GoogleModelGardenListResponse>;
    validateGoogleTarget(
      input: GoogleTargetValidationInput,
    ): Promise<GoogleTargetValidationResult>;
    deployGoogleModel(
      input: GoogleDeploymentInput,
    ): Promise<GoogleDeploymentResult>;
    saveAiProfile(
      input: DesktopAiProfileSaveInput,
    ): Promise<DesktopSettingsStatus>;
    deleteAiProfile(profileId: string): Promise<DesktopSettingsStatus>;
    saveAiConfig(input: DesktopAiConfigInput): Promise<DesktopSettingsStatus>;
    saveRepoSettings(
      input: DesktopRepoSettingsSaveInput,
    ): Promise<DesktopRepoSettings>;
  };
  health: {
    getStatus(): Promise<DesktopHealthStatus>;
  };
  review: {
    onEvent(listener: (event: ReviewRuntimeEvent) => void): () => void;
  };
  repoSync: {
    onEvent(listener: (event: RepoSyncProgressEvent) => void): () => void;
  };
}

declare global {
  interface Window {
    gitOdysseyDesktop?: GitOdysseyDesktopBridge;
  }
}

export {};

File: backend/src/api/routers/desktop.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import or_
from sqlalchemy.orm import Session

from api.api_model import (
    AIRuntimeValidationRequest,
    GoogleDeploymentRequest,
    GoogleDeploymentResponse,
    GoogleModelGardenEntryResponse,
    GoogleModelGardenListResponse,
    GoogleTargetValidationRequest,
    GoogleTargetValidationResponse,
)
from api.dependencies import get_session, get_settings
from infrastructure.ai_clients import GoogleVertexRegistry, check_adc_status
from infrastructure.ai_runtime import (
    AST_ENABLED_LANGUAGES,
    AST_SCHEMA_VERSION,
    DOCUMENT_SCHEMA_VERSION,
    AIRuntimeConfig,
    CapabilityName,
    GoogleAITarget,
    compute_embedding_fingerprint,
    describe_capability,
    load_ai_runtime_config,
    load_ai_secret_values,
)
from infrastructure.errors import AIConfigurationError, AIError
from infrastructure.settings import Settings
from data.schema import SQLEmbeddingProfile, SQLRepo

router = APIRouter()


def _get_active_embedding_fingerprint(config) -> str | None:
    if config is None or config.capabilities.embeddings is None:
        return None

    target = config.capabilities.embeddings
    return compute_embedding_fingerprint(
        target_kind=target.target_kind,
        resource_name=target.resource_name,
        project_id=config.google_project_id or "",
        location=target.location or config.google_location,
        adapter_family=target.adapter_family,
        observed_dimension=target.embedding_output_dimension,
        document_schema_version=DOCUMENT_SCHEMA_VERSION,
        ast_schema_version=AST_SCHEMA_VERSION,
        ast_enabled_languages=AST_ENABLED_LANGUAGES,
    )


def _has_reindex_required_repos(session: Session, config) -> bool:
    active_fingerprint = _get_active_embedding_fingerprint(config)
    if active_fingerprint is None:
        return False

    return (
        session.query(SQLRepo.path)
        .outerjoin(SQLRepo.embedding_profile)
        .filter(
            or_(
                SQLRepo.embedding_profile_id.is_(None),
                SQLEmbeddingProfile.fingerprint != active_fingerprint,
            )
        )
        .first()
        is not None
    )


@router.get("/health")
async def desktop_health(
    session: Session = Depends(get_session),
    settings: Settings = Depends(get_settings),
):
    runtime_error = None
    config = None
    secret_values: dict[str, str] = {}

    try:
        config = load_ai_runtime_config(settings)
        secret_values = load_ai_secret_values(settings)
    except AIConfigurationError as exc:
        runtime_error = str(exc)

    text_status = describe_capability(config, secret_values, "text_generation")
    embeddings_status = describe_capability(config, secret_values, "embeddings")
    review_status = describe_capability(config, secret_values, "review")
    embeddings_status["reindex_required"] = _has_reindex_required_repos(session, config)
    adc_status = check_adc_status()

    if runtime_error:
        text_status["message"] = runtime_error
        embeddings_status["message"] = runtime_error
        review_status["message"] = runtime_error

    return {
        "mode": "desktop",
        "authentication": {
            "desktop_backend_reachable": True,
            "desktop_user_available": True,
        },
        "ai": {
            "google": {
                "project_id": config.google_project_id if config else None,
                "location": config.google_location if config else None,
                "adc_ready": adc_status.ready,
                "adc_project_id": adc_status.project_id,
                "message": adc_status.message,
            },
            "text_generation": text_status,
            "embeddings": embeddings_status,
            "review": review_status,
        },
        "desktop_user": {
            "id": settings.desktop_user_id,
            "username": settings.desktop_user_username,
            "email": settings.desktop_user_email,
        },
    }


@router.post("/validate-ai-config")
async def validate_ai_config(
    request: AIRuntimeValidationRequest,
    session: Session = Depends(get_session),
):
    config = request.config
    secret_values = request.secret_values
    registry = GoogleVertexRegistry(config=config) if config.google_project_id else None

    text_status = describe_capability(config, secret_values, "text_generation")
    embeddings_status = describe_capability(config, secret_values, "embeddings")
    review_status = describe_capability(config, secret_values, "review")
    embeddings_status["reindex_required"] = _has_reindex_required_repos(session, config)

    for capability_name, status in (
        ("text_generation", text_status),
        ("embeddings", embeddings_status),
        ("review", review_status),
    ):
        target = getattr(config.capabilities, capability_name)
        if target is None or not status["ready"] or registry is None:
            continue
        try:
            validation = registry.validate_target(
                capability=capability_name,
                target=target,
            )
            status["ready"] = bool(validation["ready"])
            status["message"] = validation.get("message")
            if validation.get("embedding_output_dimension"):
                status["embedding_output_dimension"] = validation[
                    "embedding_output_dimension"
                ]
        except AIError as exc:
            status["ready"] = False
            status["message"] = str(exc)

    return {
        "text_generation": text_status,
        "embeddings": embeddings_status,
        "review": review_status,
    }


def _registry_for_config(config: AIRuntimeConfig) -> GoogleVertexRegistry:
    if not config.google_project_id:
        raise HTTPException(
            status_code=400,
            detail="Google Cloud project ID is required.",
        )
    return GoogleVertexRegistry(config=config)


def _entry_response(entry) -> GoogleModelGardenEntryResponse:
    return GoogleModelGardenEntryResponse(
        id=entry.id,
        resource_name=entry.resource_name,
        display_name=entry.display_name,
        publisher=entry.publisher,
        version=entry.version,
        location=entry.location,
        target_kind=entry.target_kind,
        source=entry.source,
        capabilities=entry.capabilities,
        adapter_family=entry.adapter_family,
        deployable=entry.deployable,
        description=entry.description,
    )


@router.get("/google/model-garden", response_model=GoogleModelGardenListResponse)
async def list_google_model_garden(
    google_project_id: str,
    google_location: str = "us-central1",
):
    config = AIRuntimeConfig(
        schema_version=2,
        google_project_id=google_project_id,
        google_location=google_location,
    )
    try:
        items = _registry_for_config(config).list_model_garden()
    except AIError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return GoogleModelGardenListResponse(items=[_entry_response(item) for item in items])


@router.post(
    "/google/validate-target",
    response_model=GoogleTargetValidationResponse,
)
async def validate_google_target(request: GoogleTargetValidationRequest):
    try:
        validation = _registry_for_config(request.config).validate_target(
            capability=request.capability,
            target=request.target,
        )
    except AIError as exc:
        return GoogleTargetValidationResponse(
            ready=False,
            capability=request.capability,
            target=request.target,
            message=str(exc),
        )

    target = request.target
    embedding_dimension = validation.get("embedding_output_dimension")
    if request.capability == "embeddings" and embedding_dimension:
        target = target.model_copy(
            update={"embedding_output_dimension": embedding_dimension}
        )

    return GoogleTargetValidationResponse(
        ready=True,
        capability=request.capability,
        target=target,
        message=validation.get("message"),
        embedding_output_dimension=embedding_dimension,
    )


@router.post("/google/deploy", response_model=GoogleDeploymentResponse)
async def deploy_google_model(request: GoogleDeploymentRequest):
    if not request.accepted_terms or not request.accepted_billing_notice:
        raise HTTPException(
            status_code=400,
            detail="Model Garden deployment requires terms and billing confirmation.",
        )

    registry = _registry_for_config(request.config)
    deployment = registry.get_deployment_service()
    deploy_request = deployment.construct_deployment_request(
        model_resource_name=request.model_resource_name,
        endpoint_resource_name=request.endpoint_resource_name,
        deployed_model_display_name=request.deployed_model_display_name,
        machine_type=request.machine_type,
        accelerator_type=request.accelerator_type,
        accelerator_count=request.accelerator_count,
        min_replica_count=request.min_replica_count,
        max_replica_count=request.max_replica_count,
    )
    try:
        response = deployment.deploy(deploy_request)
    except AIError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

    return GoogleDeploymentResponse(
        operation_name=response.get("name"),
        endpoint_resource_name=request.endpoint_resource_name,
        request=deploy_request,
        response=response,
    )


@router.get("/google/operations")
async def get_google_operation(
    google_project_id: str,
    google_location: str,
    operation_name: str,
):
    config = AIRuntimeConfig(
        schema_version=2,
        google_project_id=google_project_id,
        google_location=google_location,
    )
    try:
        return (
            _registry_for_config(config)
            .get_deployment_service()
            .poll_operation(operation_name)
        )
    except AIError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
